src/output/create_json.cpp

```diff
@@ -0,0 +1,184 @@
+  std::map<std::string, ProfileEntry> functions_by_paradigm;
+  std::map<std::string, ProfileEntry> messages_by_paradigm;
+  std::map<std::string, ProfileEntry> collops_by_paradigm;
+  uint64_t parallel_region_time;
+  uint64_t serial_time;
+  uint64_t num_functions;
+  uint64_t num_invocations;
+  template <typename Writer>
+  void WriteProfile(Writer& w) const;
+  WorkflowProfile() :
+    job_id(0),
+    node_count(0),
+    process_count(0),
+    thread_count(0),
+    parallel_region_time(0),
+    serial_time(0),
+    num_functions(0),
+    num_invocations(0)
+  {}
+};
+
+template <typename Map, typename Writer>
+void WriteMapUnderKey(std::string key, const Map& m, Writer& w) {
+  if(m.empty()) return;
+  w.Key(StringRef(key.c_str()));
+  w.StartObject();
+  for(auto kv : m) {
+    w.Key(StringRef(kv.first.c_str()));
+    kv.second.WriteProfile(w);
+  }
+  w.EndObject();
+}
+
+template <typename Writer>
+void WorkflowProfile::WriteProfile(Writer& w) const
+{
+  w.StartObject();
+  w.Key("job id");
+  w.Uint64(job_id);
+  w.Key("node count");
+  w.Uint(node_count);
+  w.Key("process count");
+  w.Uint(process_count);
+  w.Key("thread count");
+  w.Uint(thread_count);
+  w.Key("hardware counters");
+  w.StartArray();
+  for(auto c : counters) {
+    w.Key(StringRef(c.first.c_str()));
+    w.Uint64(c.second);
+  }
+  w.EndArray();
+  WriteMapUnderKey("functions", functions_by_paradigm, w);
+  WriteMapUnderKey("messages", messages_by_paradigm, w);
+  WriteMapUnderKey("collective operations", collops_by_paradigm, w);
+  w.Key("parallel region time"); w.Uint64(parallel_region_time);
+  w.Key("serial time"); w.Uint64(serial_time);
+  w.Key("total functions"); w.Uint64(num_functions);
+  w.Key("total calls"); w.Uint64(num_invocations);
+  w.EndObject();
+}
+
+
+bool CreateJSON(AllData& alldata) {
+
+  cout << "Creating JSON profile" << std::endl;
+  WorkflowProfile profile;
+  StringBuffer b;
+  /*Pretty*/Writer<StringBuffer> w(b);
+  for(auto n : alldata.definitions.system_tree) {
+    switch(n.data.class_id) {
+    case definitions::SystemClass::LOCATION:
+      profile.thread_count++;
+      break;
+    case definitions::SystemClass::LOCATION_GROUP:
+      profile.process_count++;
+      break;
+    case definitions::SystemClass::NODE:
+      profile.node_count++;
+      break;
+    default:
+      break;
+    }
+  }
+  static std::string bytestr("bytes");
+  static std::string timestr("time");
+  static std::string countstr("count");
+  
+  for(auto call_node : alldata.call_path_tree) {
+    auto r = alldata.definitions.regions.get(call_node.function_id);
+    if(!r) {
+      continue;
+    }
+    auto p = alldata.definitions.paradigms.get(r->paradigm_id);
+    std::string paradigm = "Compiler";
+    if(p) {
+      paradigm = p->name;
+    }
+    cout << r->name << ": " << paradigm << std::endl;
+    profile.num_functions++;
+    uint64_t excl_time = 0;
+    
+    for(auto one_node_data : call_node.node_data) {
```

Comment:
Please use const & here. Using all elements per value is quit expensive.

Reply:
Fixed in all three places you pointed out, and in other read-only iterations besides for consistency.